samples/compiler_plugins/xnnpack/src/xnnpack_sample/Transforms/LegalizeXnnpack.cpp

```diff
@@ -29,16 +29,66 @@ static FailureOr<func::FuncOp> createFuncOp(
         bodyBuild) {
   auto func = rewriter.create<func::FuncOp>(loc, name, type);
   auto *entryBlock = func.addEntryBlock();
-  auto entryBuilder = OpBuilder::atBlockBegin(entryBlock);
-  IRRewriter entryRewriter(entryBuilder);
-  if (failed(bodyBuild(entryRewriter, loc, entryBlock->getArguments(),
-                       type.getResults()))) {
-    return rewriter.notifyMatchFailure(func,
-                                       "unable to create body of function");
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(entryBlock);
+    if (failed(bodyBuild(rewriter, loc, entryBlock->getArguments(),
+                         type.getResults()))) {
+      // `bodyBuild` can leave the `FuncOp` in an invalid state, so erase it
+      // before returning.
+      rewriter.eraseOp(func);
+      return failure();
+    }
   }
   return func;
 }
 
+// TODO(ramiro050): this can be done a bit smarter without so much dynamic
+// casting. One option is to have a separate pass that attaches the result
+// dimension information as an attribute to each op.
+// TODO(ramiro050): improve name of `funcOpOperands`
+static FailureOr<SmallVector<SmallVector<Value>>> getInputOutputDims(
```

Comment:
Add a description about the function and why it is needed.

Reply:
Done